### DISCUSSION TITLE
René's Substr and performance thoughts

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#72] [Ideas]
I did some performance testing with substr, especially the looping through the string (append: 

`loop:`
`/* append would be here */` 
  ` inc r1                /* r1 loop counter */`
  ` igt r3,r1,r2        /* r2 maximum      */`
   `brf loop,r3       /* loop if false        */`
basically it's looping through the string using these 3 instructions.

if we pack them in one instruction, the code would look like this:
`loop:`
 `  incclgo r1,r2,loop`

I compared both versions with a 1,000,000,000 loop:
the original version takes about 11 seconds
the second version, 5 seconds, so we have more than 50% improvement. 
``
Nevertheless, the question is, is it worth it? What do you think?

**Comment by rvjansen** [comment]
I must say yesterday I wanted to add that I did miss instructions like BCT, BCTR, BXH and BXLE while putting these together, and even BE, BNE, but it was late already.  These would probably also save some instruction execution time while simplifying what rxc needs to generate.  So yes, it would not only be the performance gains, it would also be the advantage of less notation to have these.

**Reply by Peter-Jacob** [reply]
haha 'BCT, BCTR, BXH and BXLE' Assembler programmers of the world, unit! I also thought of BCT type instructions, but I believe counting to zero is in many cases not useful, re-calculating a string position would require an additional instruction - maybe I am wrong. Maybe we should discuss and specify them, the implementation is mostly a 3-liner.

**Reply by rvjansen** [reply]
"haha 'BCT, BCTR, BXH and BXLE' Assembler programmers of the world, unite!" - it will only make it easier to run our chains! ;-)

**Comment by adesutherland** [comment]
1. Totally Agree - these are what I call super-instructions - combining a few into one for common use cases. We just need to make sure we only create instructions that will be used! And the priority ones are ones used in loops for sure.,

For the compiler the structure of a loop is
````
Set up loop counter
loop_start:
jump to loop_end if the TO condition fails

Loop instructions

increment counter
jump to loop_start
loop_end:
````
So, for example, BGT, BLT would be used to breakout of a do loop (i.e. to loop_end), or to branch to the beginning of the loop in the case of a DO... WHILE.

But INC_BGT/INCCLGO (or whatever is a good name!) is problematic for the compiler because the structure has the increment at the end followed by an unconditional branch; However I could use a INC_BR which increments a register and does a unconditional jump? 

BTW when (in some mythical future - we are getting there) the compiler unwinds loops we would get 

````
Set up loop counter
loop_start:
jump to loop_end if the TO condition fails

Loop instructions
increment counter (only needed if the Loop Instructions use the loop counter - otherwise the counter is updated at the end with counter+=4)
Loop instructions
increment counter
Loop instructions
increment counter
Loop instructions
increment counter

jump to loop_start
loop_end:
````

2. The other key performance improvement is replacing CONCAT R1, R1, R2 with APPENDSTRING R1, R2 because CONCAT is obliged to put the result of R1 || R2 into a buffer and then move the buffer to R1. APPENDSTRING would be faster, and moreover if we then made a rule that for CONCAT Rn, Ra, Rb  n<>a and n<>b then CONCAT could be made faster too!

3. Finally, I would remove the call to the length() function. It is only called once, and you have removed error checking, but you might as well inline it. call has some overhead and it doesn't matter if you have a seperate length bif that is not used by other bifs ...

4. Finally finally, we need a signal or error instruction to allow the functions to raise the right error condition (and later still we will need to annotate the Assembler code with source code line numbers so that the REXX line number can be displayed by it - later!)

**Reply by Peter-Jacob** [reply]
sounds as we need a tech meeting and I mean an in-depth tech meeting. I think everyone has improvements in mind, let's create a plan ( I don't favour 5-years roadmaps)

**Reply by rvjansen** [reply]
we can have a tech meeting at 18:00 CET today? It *is* a cRExx Thursday.

**Reply by Peter-Jacob** [reply]
I can join!

**Reply by rvjansen** [reply]
we have a quorum - I'll send out the invite

**Reply by rvjansen** [reply]
do we favour 18:00 or 17:00 CET today?

**Reply by adesutherland** [reply]
17:00 works for me

**Reply by Peter-Jacob** [reply]
5 pm is perfect
